README.md
# NextCloudPi [![chatroom icon](https://patrolavia.github.io/telegram-badge/chat.png)](https://t.me/NextCloudPi) [![forums icon](https://img.shields.io/badge/help-forums-blue.svg)](https://help.nextcloud.com/c/support/appliances-docker-snappy-vm) [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=N8PJHSEQF4G7Y&lc=US&item_name=Own%20Your%20Bits&item_number=NextCloudPi&no_note=1&no_shipping=1&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donate_LG%2egif%3aNonHosted) [![blog](https://img.shields.io/badge/follow-blog-orange.svg)](https://ownyourbits.com)


![NC Logo](https://ownyourbits.com/wp-content/uploads/2017/11/ncp-square.png)

This is the build code for [NextCloudPi](https://ownyourbits.com/2017/02/13/nextcloud-ready-raspberry-pi-image/).

NextCloudPi is a ready to use image for Raspberry Pi, Odroid HC1, rock64 and other boards.

This code also generates the [NextCloudPi docker images](https://hub.docker.com/r/ownyourbits/nextcloudpi/) for ARM and x86 platforms, and includes an installer for any Debian based system.

## Features

 * Raspbian 9 stretch
 * Nextcloud 13.0.4
 * Apache 2.4.25, with HTTP2 enabled
 * PHP 7.2 (double the speed of PHP5!)
 * MariaDB 10
 * Redis memory cache ( NEW 11-12-2017 )
 * 4.9 Linux Kernel ( NEW 03-13-2017 )
 * ncp-config for easy setup ( RAM logs, USB drive and more )
 * Automatic redirection to HTTPS
 * ACPU PHP cache
 * PHP Zend OPcache enabled with file cache
 * HSTS
 * Cron jobs for Nextcloud
 * Sane configuration defaults
 * Full emoji support ( NEW 05-24-2017 )
 * Postfix email
 * Secure

## Extras

 * Setup wizard ( NEW 09-27-2017 )
 * NextCloudPi Web Panel ( NEW 07-24-2017 )
 * Wi-Fi ready ( NEW 03-31-2017 )
 * Ram logs ( NEW 03-31-2017 )
 * Automatic security updates, activated by default. ( NEW 03-21-2017 )
 * Let’s Encrypt for trusted HTTPS certificates.(  NEW 03-16-2017 )
 * Fail2Ban protection against brute force attacks. ( NEW 02-24-2017 )
 * UFW firewall ( NEW 07-02-2018 )
 * Dynamic DNS support for no-ip.org ( NEW 03-05-2017 )
 * Dynamic DNS support for freeDNS ( NEW 09-05-2017 )
 * Dynamic DNS support for duckDNS ( NEW 09-27-2017 )
 * Dynamic DNS support for spDYN ( NEW 11-12-2017 )
 * dnsmasq DNS server with DNS cache ( NEW 03-09-2017 )
 * ModSecurity Web Application Firewall ( NEW 03-23-2017 )
 * NFS ready to mount your files over LAN ( NEW 04-13-2017 )
 * SAMBA ready to share your files with Windows/Mac/Linux ( NEW 04-16-2017 )
 * USB automount ( NEW 05-24-2017 )
 * Remote updates ( NEW 03-31-2017 )
 * Autoumatic NCP updates ( NEW 08-16-2017 )
 * Automatic Nextcloud updates ( NEW 05-29-2018 )
 * Update notifications ( NEW 08-16-2017 )
 * NextCloud backup and restore ( NEW 05-24-2017 )
 * NextCloud online installation ( NEW 05-24-2017 )
 * Format USB drive to BTRFS ( NEW 07-03-2017 )
 * BTRFS snapshots ( NEW 04-12-2017 )
 * Automatic BTRFS snapshots ( NEW 07-02-2018 )
 * BTRFS snapshot auto sync ( NEW 19-03-2018 )
 * scheduled rsync ( NEW 19-03-2018 )
 * ZRAM ( NEW 19-03-2018 )
 * UPnP automatic port forwarding ( NEW 07-03-2017 )
 * Security audits with Lynis and Debsecan ( NEW 07-02-2018 )
 * ZRAM ( NEW 19-03-2018 )

Extras can be activated and configured using the web interface at HTTPS port 4443


![NCP-config](https://ownyourbits.com/wp-content/uploads/2017/07/web-letsencrypt.jpg)

, or from the command line from

```
sudo ncp-config
```

![NCP-config](https://ownyourbits.com/wp-content/uploads/2017/03/ncp-conf-700x456.jpg)


## How to build

Install git, docker, qemu-user-static, chroot and all the usual building tools.

```
git clone https://github.com/nextcloud/nextcloudpi.git
cd nextcloudpi
./build-SD-rpi.sh
```

, or for an Armbian based board

```
./build-SD-armbian.sh odroidxu4   # supported board code name
```

The docker armhf image can be generated in an ARM environment with docker

```
git clone https://github.com/nextcloud/nextcloudpi.git
cd nextcloudpi
make
```

, and for an x86 image, on a x86 environment do

```
git clone https://github.com/nextcloud/nextcloudpi.git
cd nextcloudpi
make nextcloudpi-x86
```

NextCloudPi can be installed in any architecture running the latest Debian

```
# curl -sSL https://raw.githubusercontent.com/nextcloud/nextcloudpi/master/install.sh | bash
```

## Downloads

Get the image, find details and more instructions at

https://ownyourbits.com/2017/02/13/nextcloud-ready-raspberry-pi-image/

Pull the docker image

https://ownyourbits.com/2017/06/08/nextcloudpi-docker-for-raspberry-pi/

https://hub.docker.com/r/ownyourbits/nextcloudpi-x86

https://hub.docker.com/r/ownyourbits/nextcloudpi-armhf

## Contact

You can find us in the [forums](https://help.nextcloud.com/c/support/appliances-docker-snappy-vm) and a [Telegram group](https://t.me/NextCloudPi)


changelog.md

[v0.57.12](https://github.com/nextcloud/nextcloudpi/commit/38fd001) (2018-06-21) nc-limits: autolimits enhancements

[v0.57.11](https://github.com/nextcloud/nextcloudpi/commit/038b260) (2018-06-20) letsencrypt: notify of renewals

[v0.57.10](https://github.com/nextcloud/nextcloudpi/commit/88843ab) (2018-06-20) ncp-web: fix JS docker detection

[v0.57.9, master](https://github.com/nextcloud/nextcloudpi/commit/ad64a4b) (2018-06-21) nc-format-USB: fix when ZRAM active

[v0.57.8 ](https://github.com/nextcloud/nextcloudpi/commit/e63523c) (2018-06-19) docker: adapt wizard

[v0.57.7 ](https://github.com/nextcloud/nextcloudpi/commit/98c99e3) (2018-06-19) docker: fix persist ncp-web password

[v0.57.6 ](https://github.com/nextcloud/nextcloudpi/commit/9d0d3b3) (2018-06-19) fix mysqld service named mysql

[v0.57.5 ](https://github.com/nextcloud/nextcloudpi/commit/e75b574) (2018-06-18) fix nextcloud-domain running before default GW is ready

[v0.57.4 ](https://github.com/nextcloud/nextcloudpi/commit/ce2d38c) (2018-06-18) letsencrypt: install from debian package

[v0.57.3 ](https://github.com/nextcloud/nextcloudpi/commit/a74248b) (2018-06-18) armbian: default to SSH disabled

[v0.57.2 ](https://github.com/nextcloud/nextcloudpi/commit/5479305) (2018-06-18) nc-static-IP: autodetect default interface

[v0.57.1 ](https://github.com/nextcloud/nextcloudpi/commit/97375a1) (2018-06-18) docker: replace systemd for service

[v0.57.0 ](https://github.com/nextcloud/nextcloudpi/commit/ef5839d) (2018-06-18) update to NC 13.0.4

[v0.56.25](https://github.com/nextcloud/nextcloudpi/commit/fe16326) (2018-06-18) nc-snapshot-sync: upgrade

[v0.56.24](https://github.com/nextcloud/nextcloudpi/commit/16ce683) (2018-06-18) nc-datadir: make sure we have the correct permissions

[v0.56.23](https://github.com/nextcloud/nextcloudpi/commit/0a047a1) (2018-06-18) nc-info: fix typo

[v0.56.22](https://github.com/nextcloud/nextcloudpi/commit/96e9f20) (2018-06-14) nc-restore: fix redis restart in docker

[v0.56.21](https://github.com/nextcloud/nextcloudpi/commit/80492b3) (2018-06-11) docker: persist SSL certificates

[v0.56.20](https://github.com/nextcloud/nextcloudpi/commit/ac4128a) (2018-06-11) ncp-web: fix sanitization for fail2ban

[v0.56.19](https://github.com/nextcloud/nextcloudpi/commit/09daa19) (2018-06-06) add ncc command, shortcut of occ

[v0.56.18](https://github.com/nextcloud/nextcloudpi/commit/8914348) (2018-06-06) NFS: fix dependency with automount

[v0.56.17](https://github.com/nextcloud/nextcloudpi/commit/b4038ca) (2018-06-04) ncp-web: sanitize the ref parameter

[v0.56.16](https://github.com/nextcloud/nextcloudpi/commit/6eaa853) (2018-05-29) build: fix cleanup armbian images

[v0.56.15](https://github.com/nextcloud/nextcloudpi/commit/6600972) (2018-05-28) ncp-web: added chinese translations

[v0.56.14](https://github.com/nextcloud/nextcloudpi/commit/c47fb2d) (2018-05-27) re-rename to NCPi

[v0.56.13](https://github.com/nextcloud/nextcloudpi/commit/b39c32f) (2018-05-28) spDYN: install curl for docker

[v0.56.12](https://github.com/nextcloud/nextcloudpi/commit/af7e6f1) (2018-05-28) nc-format-USB: fix when ZRAM active

[v0.56.11](https://github.com/nextcloud/nextcloudpi/commit/d84d292) (2018-05-27) ncp-config: silence connectivity errors

[v0.56.10](https://github.com/nextcloud/nextcloudpi/commit/a629e63) (2018-05-25) nc-ramlogs: fix docker installation from latest upstream changes

[v0.56.9 ](https://github.com/nextcloud/nextcloudpi/commit/dc41bef) (2018-05-24) remove old systemd timer config in running systems

[v0.56.8 ](https://github.com/nextcloud/nextcloudpi/commit/cf42bf8) (2018-05-22) fix php cli tmpdir for running instances

[v0.56.7 ](https://github.com/nextcloud/nextcloudpi/commit/54da658) (2018-05-22) move NC httpd logs to /var/log

[v0.56.6 ](https://github.com/nextcloud/nextcloudpi/commit/7afb720) (2018-05-20) fix update httpd log location in virtual host after nc-datadir

[v0.56.5 ](https://github.com/nextcloud/nextcloudpi/commit/d54b243) (2018-05-20) ncp-autoupdate: dont return 0 if no updates available

[v0.56.4 ](https://github.com/nextcloud/nextcloudpi/commit/656d2d8) (2018-05-18) nc-info: change port checker providers

[v0.56.3 ](https://github.com/nextcloud/nextcloudpi/commit/31543e6) (2018-05-17) nc-update-nextcloud: make sure backup syncs to disk

[v0.56.2 ](https://github.com/nextcloud/nextcloudpi/commit/5ff8675) (2018-05-15) nc-restore: refuse to restore from /var/www/nextcloud

[v0.56.1 ](https://github.com/nextcloud/nextcloudpi/commit/06f1d5c) (2018-05-15) nc-update-nextcloud: rollback in case of power cut

[v0.56.0 ](https://github.com/nextcloud/nextcloudpi/commit/0248ba3) (2018-05-15) added nc-autoupdate-nc

[v0.55.4 ](https://github.com/nextcloud/nextcloudpi/commit/ef88be9) (2018-05-15) nc-autoupdate-ncp: fix wrong user

[v0.55.3 ](https://github.com/nextcloud/nextcloudpi/commit/a57c9f6) (2018-05-15) nc-update-netcloud: include version in backup name

[v0.55.2 ](https://github.com/nextcloud/nextcloudpi/commit/e36a214) (2018-05-15) nc-backup: faster free space calculation. Minimize maintenance mode time

[v0.55.1 ](https://github.com/nextcloud/nextcloudpi/commit/a0a1145) (2018-05-14) nc-backup: exclude ncp-update-nc backups

[v0.55.0 ](https://github.com/nextcloud/nextcloudpi/commit/935cf30) (2018-05-11) added nc-update-nextcloud

[v0.54.14](https://github.com/nextcloud/nextcloudpi/commit/adf38df) (2018-05-14) samba: fix permissions

[v0.54.13](https://github.com/nextcloud/nextcloudpi/commit/570479d) (2018-05-11) nc-nextcloud: fix upload tmp dir

[v0.54.12](https://github.com/nextcloud/nextcloudpi/commit/03e9682) (2018-05-11) nc-restore: fix tmp dirs in backups without data

[v0.54.11](https://github.com/nextcloud/nextcloudpi/commit/98d34a4) (2018-05-11) nc-backup: make more robust to unexpected failure

[v0.54.10](https://github.com/nextcloud/nextcloudpi/commit/41b35c8) (2018-05-11) nc-restore: make more robust to unexpected failure


lamp.sh
#!/bin/bash

# Nextcloud LAMP base installation on Raspbian 
#
# Copyleft 2017 by Ignacio Nunez Hernanz <nacho _a_t_ ownyourbits _d_o_t_ com>
# GPL licensed (see end of file) * Use at your own risk!
#
# Usage:
# 
#   ./installer.sh lamp.sh <IP> (<img>)
#
# See installer.sh instructions for details
#
# Notes:
#   Upon each necessary restart, the system will cut the SSH session, therefore
#   it is required to save the state of the installation. See variable $STATE_FILE
#   It will be necessary to invoke this a number of times for a complete installation
#
# More at https://ownyourbits.com/2017/02/13/nextcloud-ready-raspberry-pi-image/
#

APTINSTALL="apt-get install -y --no-install-recommends"
export DEBIAN_FRONTEND=noninteractive

install()
{
  # TODO buster packages -> no php7.2-redis
    #echo "deb http://raspbian.raspberrypi.org/raspbian/ buster main contrib non-free rpi" > /etc/apt/sources.list.d/ncp-buster.list
#cat > /etc/apt/preferences <<EOF
#Package: *
#Pin: release n=stretch
#Pin-Priority: 600
#EOF

# TODO sury packages -> no php7.2-redis in ARM
echo "deb https://packages.sury.org/php/ $(lsb_release -sc) main" > /etc/apt/sources.list.d/php.list
wget -O /etc/apt/trusted.gpg.d/php.gpg https://packages.sury.org/php/apt.gpg

  apt-get update

  $APTINSTALL php7.2-redis # TODO
  exit


    # INSTALL 
    ##########################################

    # apt-get update # TODO
    $APTINSTALL apt-utils cron
    $APTINSTALL apache2
    $APTINSTALL php7.2 php7.2-curl php7.2-gd php7.2-fpm php7.2-cli php7.2-opcache \
                php7.2-mbstring php7.2-xml php7.2-zip php7.2-fileinfo php7.2-ldap \
                php7.2-intl libmagickcore-6.q16-2-extra php7.2-imagick #php-mcrypt

    # TODO
    #php-mcrypt -> only 7.1 available
    #https://github.com/nextcloud/nextcloud-snap/issues/617
    #https://github.com/onelogin/php-saml/issues/277
    #https://github.com/nextcloud/nextcloud-snap/issues/617

    mkdir -p /run/php

    # Randomize mariaDB password
    # Suggested by @enoch85 and taken from the nextcloud vm ( https://github.com/nextcloud/vm/blob/master/lib.sh#L46 )
    local DBPASSWD="default"
    echo -e "[client]\npassword=$DBPASSWD" > /root/.my.cnf
    chmod 600 /root/.my.cnf

    debconf-set-selections <<< "mariadb-server-5.5 mysql-server/root_password password $DBPASSWD"
    debconf-set-selections <<< "mariadb-server-5.5 mysql-server/root_password_again password $DBPASSWD"
    $APTINSTALL mariadb-server php7.2-mysql 
    mkdir -p /run/mysqld
    chown mysql /run/mysqld

    # CONFIGURE APACHE 
    ##########################################

  cat >/etc/apache2/conf-available/http2.conf <<EOF
Protocols h2 h2c http/1.1

# HTTP2 configuration
H2Push          on
H2PushPriority  *                       after
H2PushPriority  text/css                before
H2PushPriority  image/jpeg              after   32
H2PushPriority  image/png               after   32
H2PushPriority  application/javascript  interleaved

# SSL/TLS Configuration
SSLProtocol all -SSLv2 -SSLv3
SSLHonorCipherOrder on
SSLCipherSuite ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA256:DHE-RSA-AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:EDH-RSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA:!DSS
SSLCompression          off
SSLSessionTickets       on

# OCSP Stapling
SSLUseStapling          on
SSLStaplingResponderTimeout 5
SSLStaplingReturnResponderErrors off
SSLStaplingCache        shmcb:/var/run/ocsp(128000)
EOF

    cat >> /etc/apache2/apache2.conf <<EOF
<IfModule mod_headers.c>
  Header always set Strict-Transport-Security "max-age=15768000; includeSubDomains; preload"
</IfModule>
EOF

    # CONFIGURE PHP7
    ##########################################

    cat > /etc/php/7.2/mods-available/opcache.ini <<EOF
zend_extension=opcache.so
opcache.enable=1
opcache.enable_cli=1
opcache.fast_shutdown=1
opcache.interned_strings_buffer=8
opcache.max_accelerated_files=10000
opcache.memory_consumption=128
opcache.save_comments=1
opcache.revalidate_freq=1
opcache.file_cache=/tmp;
EOF

    a2enmod http2
    a2enconf http2 
    a2enmod proxy_fcgi setenvif
    a2enconf php7.2-fpm
    a2enmod rewrite
    a2enmod headers
    a2enmod dir
    a2enmod mime
    a2enmod ssl
    
    echo "ServerName localhost" >> /etc/apache2/apache2.conf


    # CONFIGURE LAMP FOR NEXTCLOUD
    ##########################################

    $APTINSTALL ssl-cert # self signed snakeoil certs

    # configure MariaDB ( UTF8 4 byte support )
    sed -i '/\[mysqld\]/ainnodb_large_prefix=on'       /etc/mysql/mariadb.conf.d/50-server.cnf 
    sed -i '/\[mysqld\]/ainnodb_file_per_table=1'      /etc/mysql/mariadb.conf.d/50-server.cnf 
    sed -i '/\[mysqld\]/ainnodb_file_format=barracuda' /etc/mysql/mariadb.conf.d/50-server.cnf

  # launch mariadb if not already running
  if ! pgrep -c mysqld &>/dev/null; then
    mysqld & 
  fi

  # wait for mariadb
  while :; do
    [[ -S /run/mysqld/mysqld.sock ]] && break
    sleep 0.5
  done

    mysql_secure_installation <<EOF
$DBPASSWD
y
$DBPASSWD
$DBPASSWD
y
y
y
y
EOF
}

configure() { :; }


# License
#
# This script is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This script is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this script; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330,
# Boston, MA  02111-1307  USA



etc/ncp-config.d/nc-nextcloud.sh
#!/bin/bash

# Nextcloud installation on Raspbian over LAMP base
#
# Copyleft 2017 by Ignacio Nunez Hernanz <nacho _a_t_ ownyourbits _d_o_t_ com>
# GPL licensed (see end of file) * Use at your own risk!
#
# More at https://ownyourbits.com/2017/02/13/nextcloud-ready-raspberry-pi-image/
#

VER_=13.0.4
BETA_=no
MAXFILESIZE_=2G
MEMORYLIMIT_=768M
MAXTRANSFERTIME_=3600
DBADMIN=ncadmin
REDIS_MEM=3gb
DESCRIPTION="Install any NextCloud version"

APTINSTALL="apt-get install -y --no-install-recommends"
export DEBIAN_FRONTEND=noninteractive

[ -d /var/www/nextcloud ] && {                        # don't show this during image build
INFOTITLE="NextCloud installation"
INFO="This new installation will cleanup current
NextCloud instance, including files and database.

You can later use nc-init to configure to NextCloudPi defaults

** perform backup before proceding **

You can use nc-backup "
}

install()
{
  # During build, this step is run before ncp.sh. Avoid executing twice
  [[ -f /usr/lib/systemd/system/nc-provisioning.service ]] && return 0

  # Optional packets for Nextcloud and Apps
  apt-get update
  $APTINSTALL lbzip2 iputils-ping
  $APTINSTALL -t buster php-smbclient

  # POSTFIX
  $APTINSTALL postfix || {
    # [armbian] workaround for bug - https://bugs.launchpad.net/ubuntu/+source/postfix/+bug/1531299
    echo "[NCP] Please, ignore the previous postfix installation error ..."
    mv /usr/bin/newaliases /
    ln -s /bin/true /usr/bin/newaliases
    $APTINSTALL postfix
    rm /usr/bin/newaliases
    mv /newaliases /usr/bin/newaliases
  }
 
  # REDIS
  $APTINSTALL -t buster redis-server php7.2-redis

  local REDIS_CONF=/etc/redis/redis.conf
  local REDISPASS="default"
  sed -i "s|# unixsocket .*|unixsocket /var/run/redis/redis.sock|" $REDIS_CONF
  sed -i "s|# unixsocketperm .*|unixsocketperm 770|"               $REDIS_CONF
  sed -i "s|# requirepass .*|requirepass $REDISPASS|"              $REDIS_CONF
  sed -i 's|# rename-command CONFIG ""|rename-command CONFIG ""|'  $REDIS_CONF
  sed -i "s|^port.*|port 0|"                                       $REDIS_CONF
  echo "maxmemory $REDIS_MEM" >> $REDIS_CONF
  echo 'vm.overcommit_memory = 1' >> /etc/sysctl.conf

  usermod -a -G redis www-data

  # refresh configuration, only needed in curl installer
  [[ ! -f /.ncp-image ]] && {
    systemctl restart redis-server
    systemctl enable  redis-server

    systemctl stop php7.2-fpm
    systemctl stop mysql
    sleep 0.5
    systemctl start php7.2-fpm
    systemctl start mysql
  }
  
  # service to randomize passwords on first boot
  mkdir -p /usr/lib/systemd/system
  cat > /usr/lib/systemd/system/nc-provisioning.service <<'EOF'
[Unit]
Description=Randomize passwords on first boot
Requires=network.target
After=mysql.service redis.service

[Service]
ExecStart=/bin/bash /usr/local/bin/ncp-provisioning.sh

[Install]
WantedBy=multi-user.target
EOF
  [[ "$DOCKERBUILD" != 1 ]] && systemctl enable nc-provisioning
  return 0
}

configure()
{
  ## IF BETA SELECTED ADD "pre" to DOWNLOAD PATH
  [[ "$BETA_" == yes ]] && local PREFIX="pre"
    
  ## DOWNLOAD AND (OVER)WRITE NEXTCLOUD
  cd /var/www/

  local URL="https://download.nextcloud.com/server/${PREFIX}releases/nextcloud-$VER_.tar.bz2"
  echo "Downloading Nextcloud $VER_..."
  wget -q "$URL" -O nextcloud.tar.bz2 || {
    echo "couldn't download $URL"
    return 1
  }
  rm -rf nextcloud

  echo "Installing  Nextcloud $VER_..."
  tar -xf nextcloud.tar.bz2
  rm nextcloud.tar.bz2

  ## CONFIGURE FILE PERMISSIONS
  local ocpath='/var/www/nextcloud'
  local htuser='www-data'
  local htgroup='www-data'
  local rootuser='root'

  printf "Creating possible missing Directories\n"
  mkdir -p $ocpath/data
  mkdir -p $ocpath/updater

  printf "chmod Files and Directories\n"
  find ${ocpath}/ -type f -print0 | xargs -0 chmod 0640
  find ${ocpath}/ -type d -print0 | xargs -0 chmod 0750

  printf "chown Directories\n"

  chown -R ${htuser}:${htgroup} ${ocpath}/
  chown -R ${htuser}:${htgroup} ${ocpath}/apps/
  chown -R ${htuser}:${htgroup} ${ocpath}/config/
  chown -R ${htuser}:${htgroup} ${ocpath}/data/
  chown -R ${htuser}:${htgroup} ${ocpath}/themes/
  chown -R ${htuser}:${htgroup} ${ocpath}/updater/

  chmod +x ${ocpath}/occ

  printf "chmod/chown .htaccess\n"
  if [ -f ${ocpath}/.htaccess ]; then
    chmod 0644 ${ocpath}/.htaccess
    chown ${htuser}:${htgroup} ${ocpath}/.htaccess
  fi
  if [ -f ${ocpath}/data/.htaccess ]; then
    chmod 0644 ${ocpath}/data/.htaccess
    chown ${htuser}:${htgroup} ${ocpath}/data/.htaccess
  fi

  # create and configure opcache dir
  OPCACHEDIR=/var/www/nextcloud/data/.opcache
  sed -i "s|^opcache.file_cache=.*|opcache.file_cache=$OPCACHEDIR|" /etc/php/7.2/mods-available/opcache.ini 
  mkdir -p $OPCACHEDIR
  chown -R www-data:www-data $OPCACHEDIR

  ## RE-CREATE DATABASE TABLE 
  # launch mariadb if not already running (for docker build)
  if ! pgrep -c mysqld &>/dev/null; then
    echo "Starting mariaDB"
    mysqld & 
  fi

  # wait for mariadb
  pgrep -x mysqld &>/dev/null || echo "mariaDB process not found"

  while :; do
    [[ -S /var/run/mysqld/mysqld.sock ]] && break
    sleep 0.5
  done

  echo "Setting up database..."

  # workaround to emulate DROP USER IF EXISTS ..;)
  local DBPASSWD=$( grep password /root/.my.cnf | sed 's|password=||' )
  mysql <<EOF
DROP DATABASE IF EXISTS nextcloud;
CREATE DATABASE nextcloud
    CHARACTER SET utf8mb4
    COLLATE utf8mb4_unicode_ci;
GRANT USAGE ON *.* TO '$DBADMIN'@'localhost' IDENTIFIED BY '$DBPASSWD';
DROP USER '$DBADMIN'@'localhost';
CREATE USER '$DBADMIN'@'localhost' IDENTIFIED BY '$DBPASSWD';
GRANT ALL PRIVILEGES ON nextcloud.* TO $DBADMIN@localhost;
EXIT
EOF

## SET APACHE VHOST
  echo "Setting up Apache..."
  cat > /etc/apache2/sites-available/nextcloud.conf <<'EOF'
<IfModule mod_ssl.c>
  <VirtualHost _default_:443>
    DocumentRoot /var/www/nextcloud
    CustomLog /var/log/apache2/nc-access.log combined
    ErrorLog  /var/log/apache2/nc-error.log
    SSLEngine on
    SSLCertificateFile      /etc/ssl/certs/ssl-cert-snakeoil.pem
    SSLCertificateKeyFile /etc/ssl/private/ssl-cert-snakeoil.key
  </VirtualHost>
  <Directory /var/www/nextcloud/>
    Options +FollowSymlinks
    AllowOverride All
    <IfModule mod_dav.c>
      Dav off
    </IfModule>
    LimitRequestBody 0
    SSLRenegBufferSize 10486000
  </Directory>
</IfModule>
EOF
  a2ensite nextcloud

  cat > /etc/apache2/sites-available/000-default.conf <<'EOF'
<VirtualHost _default_:80>
  DocumentRoot /var/www/nextcloud
  <IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteCond %{HTTPS} !=on
    RewriteRule ^/?(.*) https://%{SERVER_NAME}/$1 [R,L]
  </IfModule>
</VirtualHost>
EOF

  # some added security
  sed -i 's|^ServerSignature .*|ServerSignature Off|' /etc/apache2/conf-enabled/security.conf
  sed -i 's|^ServerTokens .*|ServerTokens Prod|'      /etc/apache2/conf-enabled/security.conf

  echo "Setting up system..."

  ## SET LIMITS
  sed -i "s/post_max_size=.*/post_max_size=$MAXFILESIZE_/"             /var/www/nextcloud/.user.ini 
  sed -i "s/upload_max_filesize=.*/upload_max_filesize=$MAXFILESIZE_/" /var/www/nextcloud/.user.ini 
  sed -i "s/memory_limit=.*/memory_limit=$MEMORYLIMIT_/"               /var/www/nextcloud/.user.ini 

  # TMP UPLOAD DIR
  local UPLOADTMPDIR=/var/www/nextcloud/data/tmp
  mkdir -p "$UPLOADTMPDIR"
  chown www-data:www-data "$UPLOADTMPDIR"
  sed -i "s|^;\?upload_tmp_dir =.*$|upload_tmp_dir = $UPLOADTMPDIR|" /etc/php/7.2/cli/php.ini
  sed -i "s|^;\?upload_tmp_dir =.*$|upload_tmp_dir = $UPLOADTMPDIR|" /etc/php/7.2/fpm/php.ini
  sed -i "s|^;\?sys_temp_dir =.*$|sys_temp_dir = $UPLOADTMPDIR|"     /etc/php/7.2/fpm/php.ini


  # slow transfers will be killed after this time
  cat >> /var/www/nextcloud/.user.ini <<< "max_execution_time=$MAXTRANSFERTIME_"
  cat >> /var/www/nextcloud/.user.ini <<< "max_input_time=$MAXTRANSFERTIME_"

  ## SET CRON
  echo "*/15  *  *  *  * php -f /var/www/nextcloud/cron.php" > /tmp/crontab_http
  crontab -u www-data /tmp/crontab_http
  rm /tmp/crontab_http
  
  echo "Don't forget to run nc-init"
}

# License
#
# This script is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This script is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this script; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330,
# Boston, MA  02111-1307  USA

